servicetalk-http-netty/src/test/java/io/servicetalk/http/netty/ProxyTunnel.java

```diff
@@ -126,45 +129,44 @@ private void handleRequest(final Socket socket, final InputStream in, final Stri
             final String host = authority.substring(0, colon);
             final int port = Integer.parseInt(authority.substring(colon + 1));
 
-            final Socket clientSocket = new Socket(host, port);
-            connectCount.incrementAndGet();
-            final OutputStream out = socket.getOutputStream();
-            out.write((protocol + " 200 Connection established\r\n\r\n").getBytes(UTF_8));
-            out.flush();
+            try (Socket clientSocket = new Socket(host, port)) {
+                connectCount.incrementAndGet();
+                final OutputStream out = socket.getOutputStream();
+                out.write((protocol + " 200 Connection established\r\n\r\n").getBytes(UTF_8));
+                out.flush();
 
-            final InputStream cin = clientSocket.getInputStream();
-            executor.submit(() -> copyStream(out, cin));
-            copyStream(clientSocket.getOutputStream(), in);
+                final InputStream cin = clientSocket.getInputStream();
+                Future<Void> f = executor.submit(() -> {
+                    copyStream(out, cin);
+                    return null;
+                });
+                copyStream(clientSocket.getOutputStream(), in);
+                f.get(); // wait for the copy of proxy client input to server output to finish copying.
+            }
         } else {
             throw new RuntimeException("Unrecognized initial line: " + initialLine);
         }
     }
 
-    private static void copyStream(final OutputStream out, final InputStream cin) {
+    private static void copyStream(final OutputStream out, final InputStream cin) throws IOException {
         try {
             int b;
             while ((b = cin.read()) >= 0) {
                 out.write(b);
             }
             out.flush();
```

Comment:
probably not needed since close() is imminent and implies flush()

Reply:
I'll leave the flush for now. the `close()` javadoc doesn't explicitly say anything about flushing it just says `Closes this output stream and releases any system resources associated with this stream.`.